sqldb/sqldb.smithy

```diff
@@ -0,0 +1,146 @@
+// sqldb.smithy
+// definition of an sql database capability contract
+//
+// Version 0.1 of this interface has the following features:
+//    Execute       - Execute sql operations (insert, update, create table, etc.)
+//                    Returns number of rows affected
+//    Fetch         - Select 0 or more rows from database
+//                    The returned result set is encoded in CBOR,
+//                    a language-neutral compact representation.
+//
+// CBOR is designed to be an extensible format that is language-neutral,
+// about 50-70% denser than JSON, and suitable for constrained
+// environments (low cpu and memory requirements). Parsers are simple to
+// write, and libraries are availble in [several languages](https://cbor.io).
+//
+// Not currently supported:
+// - transactions
+// - batch operations (multiple execute or fetch queries in single rpc call)
+// - streaming results
+// - prepared statements
+// - results with NULL values, array column types, or custom column data types
+//
+
+metadata package = [
+    {
+        namespace: "org.wasmcloud.interface.sqldb",
+        crate: "wasmcloud-interface-sqldb"
+     }
+]
+
+namespace org.wasmcloud.interface.sqldb
+
+use org.wasmcloud.model#wasmbus
+use org.wasmcloud.model#U32
+use org.wasmcloud.model#U64
+use org.wasmcloud.model#I64
+//use org.wasmcloud.model#codegenRust
+
+/// SqlDb - SQL Database connections
+/// To use this capability, the actor must be linked
+/// with "wasmcloud:sqldb"
+@wasmbus(
+    contractId: "wasmcloud:sqldb",
+    providerReceive: true )
+service SqlDb {
+    version: "0.1",
+    operations: [ Execute, Fetch ],
+}
+
+operation Execute {
+    input: Query,
+    output: ExecuteResult,
+}
+
+@length(min:1)
+string Query
+
+structure ExecuteResult {
+    /// optional error information.
+    /// If error is included in the FetchResult, other values should be ignored.
+    error: SqlDbError,
+
+    /// number of rows affected by the query
+    @required
+    rowsAffected: U64,
+}
+
+/// perform select query on database, returning all result rows
+operation Fetch {
+    input: Query,
+    output: FetchResult
+}
+
+
+/// Result of a fetch query
+structure FetchResult {
+    /// optional error information.
+    /// If error is included in the FetchResult, other values should be ignored.
+    error: SqlDbError,
+
+    /// number of rows returned
+    @required
+    numRows: U64,
+
+    /// description of columns returned
+    @required
+    columns: Columns,
+
+    /// result rows
+    @required
+    rows: Blob,
+}
+
+structure SqlDbError {
+
+    /// Type of error.
+    /// The list of error codes below may be expanded in the future
+    /// to provide finer-granularity failure information
+    @enum([
```

Comment:
Ooh I like that we're using the smithy-based enum here. Sets a good example for future model designers.

Reply:
There's a backlog issue to add enums to the code generator if they're specified in smithy. Currently you would call this a "documentation trait" - i.e., communication to the developer but the constraint is not technically enforced.